nn/conformer.py

```diff
@@ -0,0 +1,204 @@
+"""
+Conformer code.
+Ref: https://arxiv.org/abs/2005.08100
+"""
+
+from typing import Tuple, List, Callable, Optional, Dict, Any, Union
+from .. import nn
+import copy
+
+
+class _PositionwiseFeedForward(nn.Module):
+  """
+  Conformer position-wise feedforward neural network layer
+      FF -> Activation -> Dropout -> FF
+  """
+
+  def __init__(self, out_dim: int, dim_ff: int, dropout: float, activation: Callable[[nn.LayerRef], nn.LayerRef]):
+    """
+    :param out_dim:
+    :param dim_ff:
+    :param dropout:
+    :param activation:
+    """
+    super().__init__()
+
+    self.dropout = dropout
+    self.activation = activation
+
+    self.linear_ff = nn.Linear(n_out=dim_ff)
+    self.linear_out = nn.Linear(n_out=out_dim)
+
+  def forward(self, inp: nn.LayerRef) -> nn.LayerRef:
+    x_ff1 = self.linear_ff(inp)
+    x_act = self.activation(x_ff1)
+    x_drop = nn.dropout(x_act, dropout=self.dropout)
+    x_ff2 = self.linear_out(x_drop)
+    return x_ff2
+
+
+class _ConformerConvBlock(nn.Module):
+  """
+  Conformer convolution block
+      FF -> GLU -> depthwise conv -> BN -> Swish -> FF
+  """
+
+  def __init__(self, out_dim: int, kernel_size: int, batch_norm_opts: Optional[Dict[str, Any]] = None):
+    """
+    :param out_dim:
+    :param kernel_size:
+    """
+    super().__init__()
+
+    self.positionwise_conv1 = nn.Linear(n_out=out_dim * 2)
+    self.depthwise_conv = nn.Conv(n_out=out_dim, filter_size=(kernel_size,), groups=out_dim, padding='same')
+    self.positionwise_conv2 = nn.Linear(n_out=out_dim)
+
+    if batch_norm_opts is None:
+      batch_norm_opts = {}
+    batch_norm_opts = batch_norm_opts.copy()
+    batch_norm_opts.setdefault('epsilon', 1e-5)
+    batch_norm_opts.setdefault('momentum', 0.1)
+    self.batch_norm = nn.BatchNorm(update_sample_only_in_training=True, delay_sample_update=True, **batch_norm_opts)
+
+  def forward(self, inp: nn.LayerRef) -> nn.LayerRef:
+    x_conv1 = self.positionwise_conv1(inp)
+    x_act = nn.glu(x_conv1)
+    x_depthwise_conv = self.depthwise_conv(x_act)
+    x_bn = self.batch_norm(x_depthwise_conv)
+    x_swish = nn.swish(x_bn)
+    x_conv2 = self.positionwise_conv2(x_swish)
+    return x_conv2
+
+
+class _ConformerConvSubsample(nn.Module):
```

Comment:
Also public:
```suggestion
class ConformerConvSubsample(nn.Module):
```

Btw, is this the same VGG style subsampling we also use in other cases? (E.g. our standard attention models.) If it is just the same, maybe we should move this out and call it `VGGSubsample` or so.

Reply:
yes this is the same as VGG style subsampling. This is not specific for conformer.